### ISSUE TITLE
DataContractCriticalHelper is incorrectly declared in reference assembly

### ISSUE THREAD

**Issue by weltkante** [dotnet/runtime#80236]
### Description

The reference assembly for `System.Runtime.Serialization.Xml` has [`DataContractCriticalHelper`](https://github.com/dotnet/runtime/blob/e461b776bff0e5a6b8c72282a9f66695a837461d/src/libraries/System.Runtime.Serialization.Xml/ref/System.Runtime.Serialization.Xml.cs#L558) in namespace `System.Runtime.Serialization.DataContracts` which is not type forwarded anywhere.

`System.Private.DataContractSerialization` has a [_nested class_ `DataContractCriticalHelper`](https://github.com/dotnet/runtime/blob/8e2a059a1f5395aa7395b268ac57e5c5c23b7528/src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs#L300) inside `System.Runtime.Serialization.DataContracts.DataContract`.

Since the type is internal I'm not sure why its in the reference assembly, might not even be observable, but I've been writing some tooling and its tripping up over this mismatch between reference assemblies vs. runtime so I have to add a special case for this type.

### Reproduction Steps

check the assemblies, e.g. in ILSpy

- the reference assembly `System.Runtime.Serialization.Xml` has `DataContractCriticalHelper` in a namespace
- the runtime assembly `System.Runtime.Serialization.Xml` has no type forward for `DataContractCriticalHelper`
- the runtime assembly `System.Private.DataContractSerialization` has an implementation but its a nested class

### Expected behavior

The reference assembly should properly map to the runtime assembly (or not include the type if its not needed)

### Actual behavior

The reference assembly defines a type that does not exist in the runtime assemblies

### Configuration

.NET 7.0.0 on x64

**Comment by StephenMolloy** [comment]
This issue centers on a subtle but important aspect of reference assembly generation and API validation for types like `DataContract` that have only internal constructors.

### The problem

- **Default constructor assumption:**  
  If a class in a reference assembly does not declare any constructors, consumers and .NET tooling will assume it has a public parameterless constructor—even if the implementation does not.
- **Mismatch risk:**  
  The implementation of `DataContract` only exposes an internal constructor that takes a `DataContractCriticalHelper` argument. If the reference assembly omits this constructor, it falsely advertises a public parameterless constructor, causing API compatibility checks to fail.

### Why the helper type is included

- To accurately match the implementation, the reference assembly must declare the internal constructor with a `DataContractCriticalHelper` parameter.
- That means the reference assembly must also declare the `DataContractCriticalHelper` type—even if it’s just an empty internal class—so the constructor’s signature is valid.

### Is this exposing internals?

- No public constructor or public API uses `DataContractCriticalHelper`. Its presence is not about public visibility, but about reference assembly correctness.
- Without these internal declarations, API validation tools would incorrectly report that `DataContract` has a public parameterless constructor, which is not the case in the actual product.

### Conclusion

- Having both the internal constructor and `DataContractCriticalHelper` present in the reference assembly is **intentional and necessary**.  
- This ensures the reference and implementation assemblies accurately reflect the true API surface, avoiding confusing or misleading consumers and tools.

**Comment by jkotas** [comment]
> To accurately match the implementation, the reference assembly must declare the internal constructor with a DataContractCriticalHelper parameter.

@StephenMolloy That's not correct. Reference assemblies have all internal constructors normalized into internal constructor with no arguments if one is needed to avoid exposing a public constructor.

It is best to auto-generate the reference assemblies where possible using instructions at https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/updating-ref-source.md . The auto-generation will take care of all these subtle details for you in a uniform way. I have run the auto-generation for System.Runtime.Serialization and submitted #118311

**Comment by weltkante** [comment]
Thanks, I'd prefer the PR over the original response of closing as intended, as the reference assembly was making up an imaginary type (so it was _not_ accurately mapping the constructor) which is really annoying to work around. Its much easier to apply the rule you explained here in my tooling:

> Reference assemblies have all internal constructors normalized into internal constructor with no arguments if one is needed to avoid exposing a public constructor.

**Comment by StephenMolloy** [comment]
Yup, I misread that situation. PR merged.
